x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/HashTests.java

```diff
@@ -0,0 +1,132 @@
+    }
+
+    private static List<TestCaseSupplier> createTestCases(String algorithm) {
+        return List.of(
+            createTestCase(algorithm, DataType.KEYWORD, DataType.KEYWORD),
+            createTestCase(algorithm, DataType.KEYWORD, DataType.TEXT),
+            createTestCase(algorithm, DataType.TEXT, DataType.KEYWORD),
+            createTestCase(algorithm, DataType.TEXT, DataType.TEXT),
+            createLiteralTestCase(algorithm, DataType.KEYWORD, DataType.KEYWORD),
+            createLiteralTestCase(algorithm, DataType.KEYWORD, DataType.TEXT),
+            createLiteralTestCase(algorithm, DataType.TEXT, DataType.KEYWORD),
+            createLiteralTestCase(algorithm, DataType.TEXT, DataType.KEYWORD)
+        );
+    }
+
+    private static TestCaseSupplier createTestCase(String algorithm, DataType algorithmType, DataType inputType) {
+        return new TestCaseSupplier(algorithm, List.of(algorithmType, inputType), () -> {
+            var input = randomAlphaOfLength(10);
+            return new TestCaseSupplier.TestCase(
+                List.of(
+                    new TestCaseSupplier.TypedData(new BytesRef(algorithm), algorithmType, "algorithm"),
+                    new TestCaseSupplier.TypedData(new BytesRef(input), inputType, "input")
+                ),
+                "HashEvaluator[algorithm=Attribute[channel=0], input=Attribute[channel=1]]",
+                DataType.KEYWORD,
+                equalTo(new BytesRef(hash(algorithm, input)))
+            );
+        });
+    }
+
+    private static TestCaseSupplier createLiteralTestCase(String algorithm, DataType algorithmType, DataType inputType) {
+        return new TestCaseSupplier(algorithm, List.of(algorithmType, inputType), () -> {
+            var input = randomAlphaOfLength(10);
+            return new TestCaseSupplier.TestCase(
+                List.of(
+                    new TestCaseSupplier.TypedData(new BytesRef(algorithm), algorithmType, "algorithm").forceLiteral(),
+                    new TestCaseSupplier.TypedData(new BytesRef(input), inputType, "input")
+                ),
+                "HashConstantEvaluator[algorithm=" + algorithm + ", input=Attribute[channel=0]]",
+                DataType.KEYWORD,
+                equalTo(new BytesRef(hash(algorithm, input)))
+            );
+        });
+    }
+
+    private static String hash(String algorithm, String input) {
+        try {
+            return HexFormat.of().formatHex(MessageDigest.getInstance(algorithm).digest(input.getBytes(StandardCharsets.UTF_8)));
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalArgumentException("Unknown algorithm: " + algorithm);
+        }
+    }
+
+    @Override
+    protected Expression build(Source source, List<Expression> args) {
+        return new Hash(source, args.get(0), args.get(1));
+    }
+
+    public void testInvalidAlgorithmLiteral() {
```

Comment:
You have a test *case* for `invalid algorithm` already. Do you want to keep this one? I might move it to another class in that case - a `HashExtraTests` or something. That way you don't run it 1233242413414 times.

Reply:
I believe they are covering different branches. One added via `cases.add` is covering non-constant evaluator.
This one here is foldable, as a result it throws from `org.elasticsearch.xpack.esql.expression.function.scalar.string.Hash#toEvaluator` as input is not a valid algotythm.
Such exception is not caught and not verified by `withWarning` nor `withFoldingException`.

Please let me know if I am missing something or if exception should not be thrown there.